.config/doom/config.org
#+TITLE: Shadk's Doom Emacs Config
#+AUTHOR: Shad Kaske
#+EMAIL: shadkaske@gmail.com
#+LANGUAGE: en
#+STARTUP: inlineimages

* Table of Contents :TOC_3_gh:
- [[#configuration][Configuration]]
  - [[#user-informaiton][User Informaiton]]
    - [[#name-and-email-address][Name and Email Address]]
  - [[#emacs-ui-settings][Emacs UI Settings]]
    - [[#no-close-confirmation][No Close Confirmation]]
    - [[#clipboard-settings][Clipboard Settings]]
    - [[#theme-settings][Theme Settings]]
    - [[#font-settings][Font Settings]]
    - [[#line-numbers][Line Numbers]]
    - [[#line-spacing][Line Spacing]]
    - [[#all-the-icons][All The Icons]]
    - [[#highlight-trailing-whitespace][Highlight Trailing Whitespace]]
  - [[#keybindinds][Keybindinds]]
    - [[#global-keybinds][Global Keybinds]]
    - [[#leader-keybinds][Leader Keybinds]]
    - [[#convenience-keybinds][Convenience Keybinds]]
  - [[#toolsapp][Tools/App]]
    - [[#mu4e][mu4e]]
    - [[#org-mode][Org-Mode]]
    - [[#neotree][NeoTree]]
  - [[#languages][Languages]]

* Configuration
** User Informaiton
*** Name and Email Address

    #+BEGIN_SRC emacs-lisp
    (setq user-full-name "Shad Kaske"
        user-mail-address "shadkaske@gmail.com")
    #+END_SRC

** Emacs UI Settings
*** No Close Confirmation
    Don't bother me will a close confirmation

    #+begin_src emacs-lisp
    (setq confirm-kill-emacs nil)
    #+end_src

*** Clipboard Settings
    I prefer to manage what gets jammed into my system clipboard thank you.

   #+begin_src emacs-lisp
    (setq select-enable-clipboard nil
        x-select-enable-clipboard-manager nil)
   #+end_src

*** Theme Settings
    Default theme, currently material, and enable bold/italic fonts for the theme

    #+begin_src emacs-lisp

    (setq doom-theme 'doom-material)
    (after! doom-themes
    (setq doom-themes-enable-bold t
            doom-themes-enable-italic t))
    #+end_src

*** Font Settings
    Use JetBrains Mono for mono spaced and Big Font. Ubuntu currently for variable. Use italics for keywords and comments

   #+begin_src emacs-lisp
    (setq doom-font (font-spec :family "JetBrainsMono Nerd Font" :size 16)
        doom-variable-pitch-font (font-spec :family "Ubuntu" :size 16)
        doom-big-font (font-spec :family "JetBrainsMono Nerd Font" :size 24))
    (custom-set-faces!
    '(font-lock-comment-face :slant italic)
    '(font-lock-keyword-face :slant italic))
   #+end_src

*** Line Numbers
    I prefer using a relative line number setting

    #+begin_src emacs-lisp
    (setq display-line-numbers-type 'relative)
    #+end_src

*** Line Spacing
    Sometimes I like to mess with the spacing on the lines, sometimes not so much

    #+begin_src emacs-lisp
    ;; (setq default-text-properties '(line-spacing 0.1  line-height 1.1))
    #+end_src

*** All The Icons

    #+BEGIN_SRC emacs-lisp
    (after! all-the-icons
        (setq all-the-icons-scale-factor 1.0))
    #+END_SRC

*** Highlight Trailing Whitespace

    #+BEGIN_SRC emacs-lisp
    (setq show-trailing-whitespace t)
    #+END_SRC

** Keybindinds
    Custom keybindings
*** Global Keybinds
*** Leader Keybinds
**** NeoTree


*** Convenience Keybinds
    Just a few bits that I like to user
**** Insert New Lines Above and Below
    #+begin_src emacs-lisp
    (map!
     (:after evil
        :m "] SPC" #'evil-motion-insert-newline-below
        :m "[ SPC" #'evil-motion-insert-newline-after))
    #+end_src
**** Simple Window Navigation
    #+BEGIN_SRC emacs-lisp
    (map!
     (:after evil
       :en "C-h"   #'evil-window-left
       :en "C-j"   #'evil-window-down
       :en "C-k"   #'evil-window-up
       :en "C-l"   #'evil-window-right))
    #+END_SRC

** Tools/App
*** mu4e
    Configure mu4e to manage my email

    #+BEGIN_SRC emacs-lisp
    (remove-hook! 'mu4e-compose-pre-hook #'org-msg-mode)
    (after! mu4e
    (setq! mu4e-maildir (expand-file-name "~/Mail/kaskshad-showplace") ; the rest of the mu4e folders are RELATIVE to this one
            mu4e-get-mail-command "mbsync -c ~/.config/mu4e/mbsyncrc -a"
            mu4e-index-update-in-background t
            mu4e-compose-signature-auto-include nil
            mu4e-use-fancy-chars t
            mu4e-view-show-addresses t
            mu4e-view-show-images t
            mu4e-compose-format-flowed t
            ;mu4e-compose-in-new-frame t
            mu4e-change-filenames-when-moving t ;; http://pragmaticemacs.com/emacs/fixing-duplicate-uid-errors-when-using-mbsync-and-mu4e/
            mu4e-maildir-shortcuts
            '( ("/Inbox" . ?i)
                ("/Archive" . ?a)
                ("/Drafts" . ?d)
                ("/Deleted Items" . ?t)
                ("/Sent Items" . ?s))

            ;; Message Formatting and sending
            message-send-mail-function 'smtpmail-send-it
            message-signature-file "~/.config/mu4e/mailsignature"
            message-citation-line-format "On %a %d %b %Y at %R, %f wrote:\n"
            message-citation-line-function 'message-insert-formatted-citation-line
            message-kill-buffer-on-exit t))

    (set-email-account! "kaskshad@showplacewood.com"
                        '((user-mail-address      . "kaskshad@showplacewood.com")
                        (user-full-name         . "Shad Kaske")
                        (smtpmail-smtp-server   . "smtp.office365.com")
                        (smtpmail-smtp-service  . 587)
                        (smtpmail-stream-type   . starttls)
                        (smtpmail-debug-info    . t)
                        (mu4e-drafts-folder     . "/Drafts")
                        (mu4e-refile-folder     . "/Archive")
                        (mu4e-sent-folder       . "/Sent Items")
                        (mu4e-trash-folder      . "/Deleted Items")
                        (mu4e-update-interval   . 300)
                        ;(mu4e-sent-messages-behavior . 'delete)
                        )
                        nil)

    (mu4e-alert-set-default-style 'libnotify)
    (add-hook 'after-init-hook #'mu4e-alert-enable-notifications)
    #+END_SRC

*** Org-Mode
My Quirky Settings for using Org to GTD
**** Path Config

#+BEGIN_SRC emacs-lisp
(setq org-directory "~/Nextcloud/org/")
(setq org-noter-notes-search-path '("~/Nextcloud/org/reference/"))
#+END_SRC

**** General Org Settings
- Hide Emphasis Markers
- Enable logging
- Disable electric-mode

#+BEGIN_SRC emacs-lisp
(after! org (setq org-hide-emphasis-markers t)
  (setq org-log-done t)
  (setq org-log-into-drawer t))

(add-hook! org-mode (electric-indent-local-mode -1))

(add-hook! org-mode :append
           #'variable-pitch-mode)
#+END_SRC

**** Agenda

#+BEGIN_SRC emacs-lisp
(use-package! org-agenda
  :ensure t
  :after org
  :config
  (org-agenda-files (list org-directory))
  (setq org-agenda-custom-commands '(("g" "Scheduled today and all NEXT items" ((agenda "" ((org-agenda-span 1))) (todo "NEXT"))))))
#+END_SRC

**** Archive Settings

#+BEGIN_SRC emacs-lisp
(after! org
  (setq org-archive-tag "archive"))
  (setq org-archive-location (concat org-directory "archive.org::* From %s"))
#+END_SRC

**** Refile Targets

#+BEGIN_SRC emacs-lisp
(after! org
  (setq org-refile-targets '((nil :maxlevel . 3)
                             (org-agenda-files :maxlevel . 3))))
#+END_SRC

*** NeoTree

   #+BEGIN_SRC emacs-lisp
    (after! neotree
        (setq neo-smart-open t
              neo-windows-fixed-size nil))
   #+END_SRC

** Languages


.config/doom/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Basic Emacs Setup
(setq user-full-name "Shad Kaske"
    user-mail-address "skaske@showplacewood.com")

;; Don't Ask to Close
(setq confirm-kill-emacs nil)

;; Disable Clipboard Selection and Manager
;; (setq select-enable-clipboard nil
;;     x-select-enable-clipboard-manager nil)

;; Theme Settings
(setq doom-theme 'doom-material)
(after! doom-themes
(setq doom-themes-enable-bold t
        doom-themes-enable-italic t))

;; Font Settings
(setq doom-font (font-spec :family "JetBrainsMono Nerd Font" :size 16)
    doom-variable-pitch-font (font-spec :family "Ubuntu" :size 16)
    doom-big-font (font-spec :family "JetBrainsMono Nerd Font" :size 24))
(custom-set-faces!
'(font-lock-comment-face :slant italic)
'(font-lock-keyword-face :slant italic))

;; Relatvie Line Numbers
(setq display-line-numbers-type 'visual)

;; Line Height / Spacing
;; (setq default-text-properties '(line-spacing 0.1  line-height 1.1))

;; All The Icons Scale
(after! all-the-icons
    (setq all-the-icons-scale-factor 1.0))

;; Show Trailing Whitespace
(setq show-trailing-whitespace t)

;; Custom Keybindings
;;;; Evil Mode Line Inserts
(map!
 (:after evil
    :m "] SPC" #'evil-motion-insert-newline-below
    :m "[ SPC" #'evil-motion-insert-newline-after))

;;;; Evil Mode Simplified Navigation
;; (map!
;;  (:after evil
;;    :en "C-h"   #'evil-window-left
;;    :en "C-j"   #'evil-window-down
;;    :en "C-k"   #'evil-window-up
;;    :en "C-l"   #'evil-window-right))

;; mu4e
(remove-hook! 'mu4e-compose-pre-hook #'org-msg-mode)
(after! mu4e
(setq! mu4e-maildir (expand-file-name "~/Mail/kaskshad-showplace") ; the rest of the mu4e folders are RELATIVE to this one
        mu4e-get-mail-command "mbsync -c ~/.config/mu4e/mbsyncrc -a"
        mu4e-index-update-in-background t
        mu4e-compose-signature-auto-include nil
        mu4e-use-fancy-chars t
        mu4e-view-show-addresses t
        mu4e-view-show-images t
        mu4e-compose-format-flowed t
        ;mu4e-compose-in-new-frame t
        mu4e-change-filenames-when-moving t ;; Setting this to true causes issues tracking refiled emails that are link from org docs
        mu4e-maildir-shortcuts
        '( ("/Inbox" . ?i)
            ("/Archive" . ?a)
            ("/Drafts" . ?d)
            ("/Deleted Items" . ?t)
            ("/Sent Items" . ?s))

        ;; Message Formatting and sending
        message-send-mail-function 'smtpmail-send-it
        message-signature-file "~/.config/mu4e/mailsignature"
        message-citation-line-format "On %a %d %b %Y at %R, %f wrote:\n"
        message-citation-line-function 'message-insert-formatted-citation-line
        message-kill-buffer-on-exit t))

(set-email-account! "kaskshad@showplacewood.com"
                    '((user-mail-address      . "skaske@showplacewood.com")
                    (user-full-name         . "Shad Kaske")
                    (smtpmail-smtp-server   . "smtp.office365.com")
                    (smtpmail-smtp-service  . 587)
                    (smtpmail-stream-type   . starttls)
                    (smtpmail-debug-info    . t)
                    (mu4e-drafts-folder     . "/Drafts")
                    (mu4e-refile-folder     . "/Archive")
                    (mu4e-sent-folder       . "/Sent Items")
                    (mu4e-trash-folder      . "/Deleted Items")
                    (mu4e-update-interval   . 300)
                    ;(mu4e-sent-messages-behavior . 'delete)
                    )
                    nil)

(mu4e-alert-set-default-style 'libnotify)
(add-hook 'after-init-hook #'mu4e-alert-enable-notifications)
(setq mu4e-compose-format-flowed t)

;; Org Mode
;;;; Doom Emacs Org Settings
(setq org-directory "~/Nextcloud/org/")
(setq org-noter-notes-search-path '("~/Nextcloud/org/reference/"))

;;;; Disable LineNumber in Org Mode
(add-hook! 'org-mode-hook #'doom-disable-line-numbers-h)

;;;; Hide Emphasis Markers
(after! org
  (setq org-hide-emphasis-markers t)
  ;;;; Turn On Logging
  (setq org-log-done t)
  (setq org-log-into-drawer t)
  ;;  Org ToDo Keywords
  (setq org-todo-keywords '((sequence
                             "TODO(t)" "NEXT(n)" "WAITING(w)" "|" "DONE(d!)" "CANC(k!)")
                            (sequence
                             "ACTIVE(a)" "PLAN(p)" "HOLD(h)" "|" "COMPLETE(c!)" )))
  ;;  Org Todo Keywords Colors
  (setq org-todo-keyword-faces '(("TODO" . "DarkCyan")
                                 ("NEXT" . "DarkOrange")
                                 ("WAITING" . "DimGray")
                                 ("ACTIVE" . "LightCoral")
                                 ("PLAN" . "BlanchedAlmond")
                                 ("HOLD" . "DimGray")))
  ;;;; Org Agenda Files
  (setq org-agenda-files '("~/Nextcloud/org/inbox.org"
                          "~/Nextcloud/org/projects.org"
                          "~/Nextcloud/org/tickler.org"
                          "~/Nextcloud/org/cal/googlecal.org"
                          "~/Nextcloud/org/cal/showplacecal.org"))

  ;;;; Org Capture Templates
  (setq org-capture-templates
        '(("t" "ToDo [Inbox]"
           entry (file "~/Nextcloud/org/inbox.org")
                  "* TODO %?\n")
          ("T" "Tickler"
           entry (file "~/Nextcloud/org/tickler.org")
                  "* TODO %i%? \nSCHEDULED: <%(org-read-date nil nil \"+1d\")>")
          ("P" "New Project"
           entry (file "~/Nextcloud/org/projects.org")
                  "* %i%? ")
          ("e" "Email [Inbox]"
           entry (file "~/Nextcloud/org/inbox.org")
           "* TODO %?\n%U\n%a\n")
          ("f" "Link File [Inbox]"
           entry (file "~/Nextcloud/org/inbox.org")
           "* TODO %?\n %A\n")))

  ;;;; Org Tag Settings
  (setq org-tags-column -100)

  (setq org-tag-alist '((:startgroup . nil)
                        ("@work" . ?w)
                        ("@home" . ?h)
                        ("@errands" . ?e)
                        ("@email" . ?@)
                        ("@dev" . ?d)
                        (:endgroup . nil)
                        ("chad" . ?c)
                        ("andy" . ?a)
                        ("travis" . ?t)))

  ;;;; Org Refiler Targets
  (setq org-refile-targets '((org-agenda-files :maxlevel . 3)))
  (advice-add 'org-refile :after 'org-save-all-org-buffers)
  (setq org-ellipsis "  ▼")
  (setq org-superstar-headline-bullets-list '("◉" "○" "●" "○" "●" "○" "●"))
  (dolist (face '((org-level-1 . 1.2)
                  (org-level-2 . 1.1)
                  (org-level-3 . 1.05)
                  (org-level-4 . 1.0)
                  (org-level-5 . 0.9)
                  (org-level-6 . 0.85)
                  (org-level-7 . 0.8)
                  (org-level-8 . 0.75)))))

;;;; Disable Electric Indent Mode
(add-hook! org-mode (electric-indent-local-mode -1))

;;;; Archiving
(use-package! org-archive
  :after org
  :config
  (setq org-archive-location "archive.org::datetree/"))

;;;; Visual Fill Column Mode
(defun smk/org-mode-visual-fill ()
  (setq visual-fill-column-width 120
        visual-fill-column-center-text t)
  (visual-fill-column-mode 1))

(use-package! visual-fill-column
  :hook (org-mode . smk/org-mode-visual-fill))

;;;; Org Appear Mode - Show emphasis markers on cursor
(add-hook! 'org-mode-hook #'org-appear-mode)
(setq org-appear-autolinks t)

;; NeoTree
(after! neotree
    (setq neo-smart-open t
          neo-windows-fixed-size nil))

;; Centaur Tabs
;; (setq centaur-tabs-set-bar 'over)
;; (setq centaur-tabs-close-button nil)

;; Dired

;; LSP Settings
(setq lsp-file-watch-threshold nil)

;; Company Mode Settings
(set-company-backend! 'org-mode '(company-yasnippet company-capf company-files company-elisp))
(set-company-backend! 'emacs-lisp-mode '(company-yasnippet company-elisp))
;; (set-company-backend! 'mu4e-compose-edit '(company-mu company-yasnippet))
(setq company-idle-delay 0.25
      company-minimum-prefix-length 3)
(add-to-list 'company-backends '(company-capf company-files company-yasnippet company-semantic company-bbdb company-cmake company-keywords))
